drivers/mtd/at25ee.c

```diff
@@ -0,0 +1,1033 @@
+ *   nblocks    - nblocks to read
+ *   buf        - pointer to variable to store the read data
+ *
+ * Returned Value:
+ *   Number of blocks written
+ ****************************************************************************/
+
+static ssize_t at25ee_bread(FAR struct mtd_dev_s *dev,
+                           off_t startblock,
+                           size_t nblocks, FAR uint8_t *buf)
+{
+  FAR struct at25ee_dev_s *priv = (FAR struct at25ee_dev_s *)dev;
+  off_t offset;
+  ssize_t nread;
+  size_t i;
+
+  DEBUGASSERT(dev);
+  DEBUGASSERT(buf);
+
+  if (priv->blocksize > priv->pgsize)
+    {
+      startblock *= (priv->blocksize / priv->pgsize);
+      nblocks    *= (priv->blocksize / priv->pgsize);
+    }
+
+  finfo("startblock: %08lx nblocks: %lu\n",
+        (unsigned long)startblock, (unsigned long)nblocks);
+
+  if (startblock >= priv->npages)
+    {
+      return 0;
+    }
+
+  if (startblock + nblocks > priv->npages)
+    {
+      nblocks = priv->npages - startblock;
+    }
+
+  /* Convert the access from startblock and number of blocks to a byte
+   * offset and number of bytes.
+   */
+
+  offset = startblock * priv->pgsize;
+
+  /* Then perform the byte-oriented read for each block separately */
+
+  for (i = 0; i < nblocks; i++)
+    {
+      nread = at25ee_read(dev, offset, priv->pgsize, buf);
+      if (nread < 0)
+        {
+          return nread;
+        }
+
+      offset += priv->pgsize;
+      buf += priv->pgsize;
+    }
+
+  if (priv->blocksize > priv->pgsize)
+    {
+      return nblocks / (priv->blocksize / priv->pgsize);
+    }
+  else
+    {
+      return nblocks;
+    }
+}
+
+/****************************************************************************
+ * Name: at25ee_bwrite
+ *
+ * Description:
+ *   Write a number of blocks of data.
+ *
+ * Input Parameters:
+ *   dev        - a reference to the device structure
+ *   startblock - starting block to write to
+ *   nblocks    - nblocks to write
+ *   buf        - pointer to data buffer to write
+ *
+ * Returned Value:
+ *   Size of the data written
+ ****************************************************************************/
+
+static ssize_t at25ee_bwrite(FAR struct mtd_dev_s *dev, off_t startblock,
+                             size_t nblocks, FAR const uint8_t *buf)
+{
+  FAR struct at25ee_dev_s *priv = (FAR struct at25ee_dev_s *)dev;
+  size_t blocksleft;
+
+  DEBUGASSERT(dev);
+  DEBUGASSERT(buf);
+
+#if 1
+  if (priv->blocksize > priv->pgsize)
+    {
+      startblock *= (priv->blocksize / priv->pgsize);
+      nblocks    *= (priv->blocksize / priv->pgsize);
+    }
+#endif
+
+  blocksleft  = nblocks;
+
+  if (startblock >= priv->npages)
+    {
+      return 0;
+    }
+
+  if (startblock + nblocks > priv->npages)
+    {
+      nblocks = priv->npages - startblock;
+    }
+
+  finfo("startblock: %08lx nblocks: %d\n", (long)startblock, (int)nblocks);
+
+  while (blocksleft-- > 0)
+    {
+      uint16_t offset = startblock * priv->pgsize;
```

Comment:
```suggestion
      off_t offset = startblock * priv->pgsize;
```

Reply:
Serves me right for using other people's code as the staring point :-) Will change